distros/P/POSIX-Run-Capture/runcap/runcap.3
.\" This file is part of runcap -*- nroff -*-
.\" Copyright (C) 2017-2024 Sergey Poznyakoff
.\"
.\" Runcap is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3, or (at your option)
.\" any later version.
.\"
.\" Runcap is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with runcap.  If not, see <http://www.gnu.org/licenses/>.
.TH RUNCAP 2 "March 14, 2024" "RUNCAP" "User Commands"
.SH NAME
runcap \- run external process and capture its stdout and stderr
.SH SYNOPSIS
.nf
.B #include <runcap.h>
.sp
.BI "int runcap(struct runcap *" rc ", int " flags );
.BI "void runcap_free(struct runcap *" rc );
.sp
.BI "int runcap_getc(struct runcap *" rc ", int " stream ", char *" cp );
.BI "ssize_t runcap_getline(struct runcap *" rc ", int " stream ,
.BI "                       char **" pstr ", size_t *" psize );
.sp
.BI "off_t runcap_tell(struct runcap *" rc ", int " stream );
.BI "off_t runcap_seek(struct runcap *" rc ", int " stream ", off_t " off ", int " whence);
.BI "int runcap_rewind(struct runcap *" rc ", int " stream );
.SH DESCRIPTION
The function
.B runcap
runs an external command, and waits for its termination, capturing
its standard output and standard error streams, and optionally
supplying data at its standard input.
.PP
The \fIrc\fR argument points to \fBstruct runcap\fR, which serves two
purposes.  On input, it supplies the necessary data for running the
command.  The caller is responsible for pointing the 
.B rc_argv
member to the vector of command line arguments.  This member must
always be initialized.  If any of the remaining members are
initialized, the caller must inform the
.B runcap
function about it by setting appropriate bit in the \fIflags\fR
argument.
.PP
The function returns 0 on success, and -1 on error.
.PP
Upon return from
.BR runcap ,
the remaining members of the structure will be filled with the
information about the termination status of the program and the
captured content of its standard output and standard error streams.
.PP
Whatever the return status, when no longer used the structure must be
freed using
.BR runcap_free .
.PP
.SS The struct runcap argument
The \fBstruct runcap\fR is defined as follows:
.PP
.in +4n
.nf
struct runcap
{
    char *rc_program;    /* [\fIIN\fR] (Path)name of the program to run. */ 
    char **rc_argv;      /* [\fIIN\fR] Argument vector. */
    char **rc_env;       /* [\fIIN\fR] Environment variables. */
    unsigned rc_timeout; /* [\fIIN\fR] Execution timeout. */
    struct stream_capture rc_cap[3]; 
    pid_t rc_pid;        /* [\fIOUT\fR] PID of the process. */
    int rc_status;       /* [\fIOUT\fR] Termination status. */
    int rc_errno;        /* [\fIOUT\fR] System error code. */
};
.fi
.in
.PP
Members that may be initialized on input are marked with
[\fIIN\fR], those that are set upon return from \fBruncap()\fR are
marked with [\fIOUT\fR].
.PP
The caller is responsible for initializing the \fBrc_argv\fR member
with a pointer to the command line arguments array.  Other
input fields are:
.TP
.B rc_program
Actual pathname of the program to be executed.  If initialized, the
\fBRCF_PROGRAM\fR bit must be set in \fIflags\fR.  Otherwise, the
value of \fBrc_argv[0]\fR is assumed.
.TP
.B rc_timeout
Time to wait for the program termination, in seconds.  If initialized,
the \fBRCF_TIMEOUT\fR bit must be set in \fIflags\fR.  If not set,
.B runcap
will wait indefinitely.
.TP
.B rc_env
A
.BR NULL -terminated
array of environment variables.  Each element (except the last
.BR NULL )
has the form "\fIname\fR=\fIvalue\fR".  If initialized, the
\fBRCF_ENV\fR bit must be set in \fIflags\fR.
.PP
The three streams associated with the running command are described by
the
.B rc_cap
array.  Its elements correspond to the standard input, output and
error streams.  Upon successful return, the captured
data from stdout and stderr can be retrieved using the \fBruncap_getc\fR, and
\fBruncap_getline\fR functions (see below).  For convenience, the
following constants are defined in \fBruncap.h\fR to refer to the
corresponding streams:
.BR RUNCAP_STDIN ,
.BR RUNCAP_STDOUT ,
.BR RUNCAP_STDERR .
.PP
The \fBstream_capture\fR structure contains the following members
accessible to the caller:
.PP
.in +4n
.nf
struct stream_capture
{
    size_t sc_size;      /* [\fIIN\fR] size of the buffer */
    void (*sc_linemon)(const char *, size_t, void *);
                         /* [\fIIN\fR] Line monitor function */
    void  *sc_monarg;    /* [\fIIN\fR] Value for its 3rd parameter */

    off_t  sc_leng;      /* [\fIOUT\fR] total length of captured data */
    size_t sc_nlines;    /* [\fIOUT\fR] number of captured lines */

    int    sc_storfd;    /* [\fIIN\fR] storage file descriptor */      

    /* The following two are available only in \fBrc_cap[RUNCAP_STDIN]\fR
       For details. see the subsection \fBSupplying standard input\fR.
     */
    int  sc_fd;          /* Input file descriptor */
    char *sc_base;       /* Input data */
};
.fi
.in
.PP
The caller may initialize the following members in
\fBrc_cap[RUNCAP_STDOUT]\fR and \fBrc_cap[RUNCAP_STDERR]\fR:
.TP
.B sc_size
Size of the buffer (\fBsize_t\fR).  If set, the \fBRCF_STDOUT_SIZE\fR
(for \fBRUNCAP_STDOUT\fR) or \fBRCF_STDERR_SIZE\fR (for
\fBRUNCAP_STDERR\fR) bit must be set in \fIflags\fR.  If the amount of
input data exceeds the buffer size, the data are saved in the disk
file, therefore setting a larger buffer can improve performance.

Setting \fBsc_size\fR to 0 disables capturing.  Same effect is
achieved by setting the \fBRCF_STDOUT_NOCAP\fR or
\fBRCF_STDERR_NOCAP\fR flag.

Default buffer size is 4096 bytes.
.TP
.B sc_linemon
A pointer to the line monitor function.  If set, the
\fBRCF_STDOUT_LINEMON\fR (for \fBRUNCAP_STDOUT\fR), or
\fBRCF_STDERR_LINEMON\fR (for \fBRUNCAP_STDERR\fR) bit must be set in
\fIflags\fR.
.sp
Line monitor function allows the caller to monitor the arrival of
the new data in the corresponding stream.  It is declared as follows:
.sp
.BI "void linemon(const char *" line ", size_t " size ", void *" data )
.sp
where \fIline\fR is the line that has been read, \fIsize\fR is its
length, and \fIdata\fR is an opaque pointer to application-specific
data, supplied in the \fBsc_monarg\fR member.
.sp
The line monitor function is invoked to notify the caller about
arrival of new data on the stream.  There are 4 cases:
.RS +8
.nr step 1 1
.IP \n[step].
A newline newline character is encountered in the stream.
.sp
The \fIline\fR argument points to the beginning of the captured
line.  The \fIsize\fR argument contains number of characters in the
line, including the newline character.
.IP \n+[step].
The stream buffer is full (and therefore is about to be flushed into
the storage file) and contains unfinished line.
.sp
The \fIline\fR
argument points to the beginning of the captured line fragment.  The
\fIsize\fR argument contains number of characters in the 
fragment.
.IP \n+[step].
No more characters are available in the stream and the
stream buffer contains unreported characters.
.sp
Argument assignment is the same as in previous case.
.IP \n+[step].
The stream reached EOF.
.sp
The \fIsize\fR argument is 0.  The
\fIline\fR argument is not meaningful, although it does point to a
valid memory location.
.PP
Consequently, if the \fBsc_linemon\fR function needs to process full
lines. it has to keep the state of processing (e.g. in the \fBdata\fR
argument), and concatenate the line parameters until the following
condition is met:
.sp
.EX
  size == 0 || line[size-1] == '\\n'
.EE  
.RE
.TP
.B sc_monarg
The value of the \fIdata\fR parameter for the \fBsc_linemon\fR
function.
.TP
.B sc_storfd
File descriptor of the storage file for that stream.  If initialized, the
\fBRCF_STDOUT_STORFD\fR or \fBRCF_STDERR_STORFD\fR bit must be set in
\fIflags\fR.  All data obtained by that stream will be written to this
descriptor, in addition to the normal processing.
.SS Supplying standard input
The
.B rc_cap[RUNCAP_STDIN]
field can be used to supply standard input for the command.  The input
can be supplied either as a character string, or as a file
descriptor.  To use the first method, initialize
.B rc_cap[RUNCAP_STDIN].sc_base
with the pointer to the string,
.B rc_size
with its length, and set
.B sc_fd
to -1.
.PP
To use the second method, set
.B rc_cap[RUNCAP_STDIN].sc_fd
to the file descriptor opened for reading, and set
.B rc_cap[RUNCAP_STDIN].sc_base
to
.BR NULL .
.PP
Whichever approach is used, set the \fBRCF_STDIN\fR bit in \fIflags\fR
to inform
.B runcap()
about the fact.
.SS Output
Upon return, the following fields are initialized:
.TP
.B rc_status
Termination status, as returned by
.BR wait (2).
.TP
.B rc_errno
Value of errno, if terminated on error.
.PP
The following fields are defined upon successful return:
.TP
.B rc_cap[RUNCAP_STDOUT].sc_leng
Total length of captured stdout.
.TP
.B rc_cap[RUNCAP_STDOUT].sc_nlines
Number of lines in the captured stdout.
.TP
.B rc_cap[RUNCAP_STDERR].sc_leng
Total length of captured stderr.
.TP
.B rc_cap[RUNCAP_STDERR].sc_nlines
Number of lines in the captured stderr.
.PP
The last captured line can be partial, i.e. not terminated with a
newline character.  It will be counted in the stream's \fBsc_nlines\fR
anyways.
.PP
Captured data can be retrieved using the \fBruncap_getc\fR, and
\fBruncap_getline\fR functions, described below.
.SS Examining output
Upon return from \fBruncap\fR the following functions can be used to
retrieve captured data from the \fBstruct runcap\fR object pointed
to by its \fIrc\fR argument.  The stream to retrieve data from is
identified by the \fIstream\fR argument, whose valid values are
\fBRUNCAP_STDOUT\fR (or \fB1\fR) or \fBRUNCAP_STDERR\fR (or \fB2\fR).
.PP
The function \fBruncap_getc\fR reads the next character from the
captured stream and returns it as an unsigned char cast to an int.
It returns 0 on end of stream, and -1 on error.  In the latter case,
the \fBerrno\fR variable contains the error code, as usual.
.PP
The function \fBruncap_getline\fR reads all characters from the
current position in the stream up to and including the next newline
character (ASCII 10).  It will allocate the buffer for the characters
as necessary and will store the address of the buffer into
\fB*pstr\fR, and its allocated size in \fB*psize\fR.  The buffer is
null-terminated and includes the newline character, if one was found.
.PP
If \fB*pstr\fR is \fBNULL\fR, the function will allocate a buffer of
sufficient size for storing the line.
.PP
Otherwise, \fB*pstr\fR should contain a pointer to a buffer
\fB*psize\fR bytes in size, allocated using
.BR malloc (3).
If the buffer is not large enough to hold the characters,
.B runcap_getline
will resize it using
.BR realloc (3),
updating \fB*pstr\fR and \fB*psize\fR as necessary.
.PP
On success,
.B runcap_getline
returns the number of characters (including the newline) stored in the
buffer, or 0 if end of stream is reached.  On error, it returns -1 and
sets \fBerrno\fR.
.PP
The function \fBruncap_tell\fR returns offset in bytes to the current
position in the requested stream.
.PP
The function \fBruncap_seek\fR repositions the offset of the requested
\fIstream\fR to the argument \fIoffset\fR according to the directive
\fIwhence\fR as follows:
.TP
.B SEEK_SET
The offset is set to \fIoffset\fR bytes.
.TP
.B SEEK_CUR
The offset is set to its current location plus \fIoffset\fR bytes.
.TP
.B SEEK_END
The offset is set to the size of the stream (\fBsc_leng\fR) plus
\fIoffset\fR bytes.
.PP
The function returns 0 on success.  On error, it returns -1 and sets
\fBerrno\fR.
.PP
The function \fBruncap_rewind\fR repositions the current offset of
\fIstream\fR to 0.
.PP
The following pairs of calls are equivalent:
.PP
.in +4n
.nf
runcap_tell(rc, stream) <=> runcap_seek(rc, stream, 0, SEEK_CUR)
.fi
.in
.sp
and
.sp
.in +4n
.nf
runcap_rewind(rc, stream) <=> runcap_seek(rc, stream, 0, SEEK_SET)
.fi
.in
.SS Redirecting output
In addition to the normal processing, the output sent to a particular
stream can be redirected to a file.  To do so, initialize the
\fBsc_storfd\fR member of the output stream to the file descriptor
and set the \fBRCF_STDOUT_STORFD\fR or \fBRCF_STDERR_STORFD\fR
flag (depending on the stream).  For example, the following code
redirects the standard output of the program to the file "outfile":
.PP
.in +4n
.nf
int fd = open("outfile", O_CREAT|O_RDWR|O_TRUNC, 0600);
assert(fd != -1);
rc->rc_cap[RUNCAP_STDOUT].sc_storfd = fd;

runcap(&rc, RCF_STDOUT_STORFD);
.fi
.in
.SS Disabling capturing
Two ways are provided to disable capturing of a particular stream.
First, you can set the \fBsc_size\fR member of the corresponding
\fBrc_cap\fR element to 0 and indicate that by setting the
\fBRCF_STDOUT_SIZE\fR (for stdout) or \fBRCF_STDERR_SIZE\fR (for
stderr) flag.  This was the method used in first releases of runcap.
.PP
Another way is to use the \fBRCF_STDOUT_NOCAP\fR and/or
\fBRCF_STDERR_NOCAP\fR flag in the call to \fBruncap\fR.
.PP
Both methods are equivalent.
.PP
When capturing of a particular stream is disabled, the corresponding
file descriptor of the program being invoked is inherited from the
calling program.  This means, e.g., that if you disable stderr
capturing, error diagnostics from the programs that \fBruncap\fR runs
will appear at the stderr of your program.  Upon return from
\fBruncap\fR both
.B sc_leng
and
.B sc_nlines
fields of the corresponding \fBrc_cap\fR structure will be zero.
.PP
This is changed if line monitor is installed on the stream with
disabled capturing.  In that case, the output from the program will be
intercepted, passed to the line monitor function and discarded.  Upon
return, both 
.B sc_leng
and
.B sc_nlines
will reflect the actual length in bytes and number of lines in the
processed material.
.B sc_nlines
field.
.SH RETURN VALUE
Upon successful completion,
.B runcap()
returns 0.  The \fBrc.rc_status\fR value should be inspected to see if
the program terminated successfully.  On error, it sets the
\fBrc.rc_errno\fR and returns -1.
.PP
The
.B runcap_getc()
returns the retrieved character on success, 0 if end of stream is hit,
and -1 on error.
.PP
The function
.B runcap_getline()
returns the number of retrieved characters (including the newline) on
success, 0 on end of stream, and -1 on error.
.PP
.B runcap_tell()
and 
.B runcap_seek()
return the current offset (a non-negative value) on success, and -1
on error.
.PP
.B runcap_rewind()
returns 0 on success and -1 on error.
.PP
When returning an error (-1), all functions set the global \fBerrno\fR
variable to the code describing the error. 
.SH EXAMPLE
The function below runs the
.BR tar (1)
command in verbose mode and returns its output as an array of
strings.
.PP
.in +4n
.nf
char **
archive(void)
{
    struct runcap rc;
    int res;
    char *av[] = { "tar", "cfv", "src.tar", "src", NULL };
    char **ret = NULL;
    
    rc.rc_program = "/bin/tar";
    rc.rc_argv = av;
    res = runcap(&rc, RCF_PROGRAM);
    if (res == -1) {

distros/P/POSIX-Run-Capture/runcap/runcap.c
/* runcap - run program and capture its output
   Copyright (C) 2017-2024 Sergey Poznyakoff

   Runcap is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 3 of the License, or (at your
   option) any later version.

   Runcap is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License along
   with Runcap. If not, see <http://www.gnu.org/licenses/>. */

#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/select.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>
#include <string.h>
#include "runcap.h"

static int
stream_capture_init(struct stream_capture *cap, size_t size, int flags)
{
	if (!cap) {
		errno = EINVAL;
		return -1;
	}

	if (size == 0) {
		if (flags & RCF_SC_LINEMON)
			size = STRCAP_BUFSIZE;
		flags |= RCF_SC_NOCAP;
	} else if ((flags & RCF_SC_NOCAP) && !(flags & RCF_SC_LINEMON))
		size = 0;
	
	if (size) {
		cap->sc_base = malloc(size);
		if (!cap->sc_base)
			return -1;
	} else
		cap->sc_base = NULL;
	cap->sc_size = size;
	cap->sc_leng = 0;
	cap->sc_level = 0;
	cap->sc_nlines = 0;
	cap->sc_cur = 0;
	if (!(flags & RCF_SC_STORFD))
		cap->sc_storfd = -1;
	cap->sc_fd = -1;

	if (!(flags & RCF_SC_LINEMON)) {
		cap->sc_linemon = NULL;
		cap->sc_monarg = NULL;
	}
	cap->sc_flags = flags;
	
	return 0;
}

static void
stream_capture_reset(struct stream_capture *cap)
{
	cap->sc_leng = 0;
	cap->sc_level = 0;
	cap->sc_nlines = 0;
	cap->sc_cur = 0;

	if (cap->sc_storfd >= 0) {
		if (cap->sc_flags & RCF_SC_STORFD) {
			/* FIXME: Restore fd position? */;
		} else {
			close(cap->sc_storfd);
			cap->sc_storfd = -1;
		}
	}
	if (cap->sc_fd >= 0) {
		close(cap->sc_fd);
		cap->sc_fd = -1;
	}
}

static void
stream_capture_free(struct stream_capture *cap)
{
	stream_capture_reset(cap);
	free(cap->sc_base);
	cap->sc_base = NULL;
	cap->sc_size = 0;

	cap->sc_linemon = NULL;
	cap->sc_monarg = NULL;
}

static int
full_write(int fd, char *buf, size_t size)
{
	while (size) {
		ssize_t n = write(fd, buf, size);
		if (n == -1)
			return -1;
		if (n == 0) {
			errno = ENOSPC;
			return -1;
		}
		buf += n;
		size -= n;
	}
	return 0;
}

static int
stream_capture_flush(struct stream_capture *cap)
{
	int res;
		
	if (cap->sc_level == 0)
		return 0;
	if (cap->sc_linemon && cap->sc_cur < cap->sc_level)
		cap->sc_linemon(cap->sc_base + cap->sc_cur,
				cap->sc_level - cap->sc_cur,
				cap->sc_monarg);
	if (!(cap->sc_flags & RCF_SC_NOCAP)) {
		if (cap->sc_storfd == -1) {
			int fd;
			char tmpl[] = "/tmp/rcXXXXXX";
			fd = mkstemp(tmpl);
			if (fd == -1)
				return -1;
			unlink(tmpl);
			cap->sc_storfd = fd;
		}
		res = full_write(cap->sc_storfd, cap->sc_base, cap->sc_level);
		if (res)
			return -1;
	}
	cap->sc_level = 0;
	cap->sc_cur = 0;
	return 0;
}

static int
stream_capture_get(struct stream_capture *cap, int *feof)
{
	int rc;
	size_t i;
	
	if (cap->sc_level == cap->sc_size) {
		if (stream_capture_flush(cap))
			return -1;
	}
		
	rc = read(cap->sc_fd, cap->sc_base + cap->sc_level,
		  cap->sc_size - cap->sc_level);
	if (rc == -1) {
	        if (errno == EINTR)
		      return 0;
		return -1;
	}
	if (rc == 0) {
		if (cap->sc_linemon) {
			if (cap->sc_level > cap->sc_cur) {
				cap->sc_linemon(cap->sc_base + cap->sc_cur,
						cap->sc_level - cap->sc_cur,
						cap->sc_monarg);
				cap->sc_cur = cap->sc_level;
				cap->sc_nlines++;
			}
			cap->sc_linemon(cap->sc_base, 0, cap->sc_monarg);
		}
		*feof = 1;
		return 0;
	} else
		*feof = 0;

	i = cap->sc_level;

	cap->sc_level += rc;
	cap->sc_leng  += rc;
	
	for (; i < cap->sc_level; i++) {
		if (cap->sc_base[i] == '\n') {
			if (cap->sc_linemon)
				cap->sc_linemon(cap->sc_base + cap->sc_cur,
						i - cap->sc_cur + 1,
						cap->sc_monarg);
			cap->sc_cur = i + 1;
			cap->sc_nlines++;
		}
	}

	if ((cap->sc_flags & RCF_SC_NOCAP)
	    && cap->sc_linemon
	    && cap->sc_level > cap->sc_cur) {
		memmove(cap->sc_base, cap->sc_base + cap->sc_cur,
			cap->sc_level - cap->sc_cur);
		cap->sc_level -= cap->sc_cur;
		cap->sc_cur = 0;
	}

	return 0;
}

static int
stream_capture_put(struct stream_capture *cap, int *feof)
{
	if (cap->sc_cur < cap->sc_level) {
		int n = write(cap->sc_fd, &cap->sc_base[cap->sc_cur], 1);
		if (n == -1) {
		        if (errno == EINTR)
			      return 0;
			return -1;
		}
		if (n == 0) {
			errno = ENOSPC;
			return -1;
		}
		cap->sc_cur++;
	}
	*feof = cap->sc_cur == cap->sc_level;
	return 0;
}

void
runcap_free(struct runcap *rc)
{
	stream_capture_free(&rc->rc_cap[RUNCAP_STDIN]);
	stream_capture_free(&rc->rc_cap[RUNCAP_STDOUT]);
	stream_capture_free(&rc->rc_cap[RUNCAP_STDERR]);
}

static inline int
timeval_after(struct timeval const *a, struct timeval const *b)
{
	if (a->tv_sec == b->tv_sec)
		return a->tv_usec < b->tv_usec;
	else
		return a->tv_sec < b->tv_sec;
}

static inline struct timeval
timeval_diff(struct timeval const *a, struct timeval const *b)
{
	struct timeval res;

	res.tv_sec = a->tv_sec - b->tv_sec;
	res.tv_usec = a->tv_usec - b->tv_usec;
	if (res.tv_usec < 0) {
		--res.tv_sec;
		res.tv_usec += 1000000;
	}

	return res;
}

extern char **environ;

static int
runcap_start(struct runcap *rc)
{
	int p[RUNCAP_NBUF][2] = { { -1, -1}, { -1, -1 }, { -1, -1 } };
	int i;
	
	for (i = 0; i < RUNCAP_NBUF; i++)
		if (rc->rc_cap[i].sc_size) {
			if (pipe(p[i])) {
				goto err;
			}
		}

	switch (rc->rc_pid = fork()) {
	case 0: /* Child */
		if (p[RUNCAP_STDIN][0] >= 0) {
			dup2(p[RUNCAP_STDIN][0], RUNCAP_STDIN);
			close(p[RUNCAP_STDIN][1]);
		} else if (rc->rc_cap[RUNCAP_STDIN].sc_fd >= 0) {
			dup2(rc->rc_cap[RUNCAP_STDIN].sc_fd, RUNCAP_STDIN);
		}

		if (p[RUNCAP_STDOUT][0] >= 0) {
			dup2(p[RUNCAP_STDOUT][1], RUNCAP_STDOUT);
			close(p[RUNCAP_STDOUT][0]);
		} else if (rc->rc_cap[RUNCAP_STDOUT].sc_storfd != -1) {
			dup2(rc->rc_cap[RUNCAP_STDOUT].sc_storfd,
			     RUNCAP_STDOUT);
		}			
		
		if (p[RUNCAP_STDERR][0] >= 0) {
			dup2(p[RUNCAP_STDERR][1], RUNCAP_STDERR);
			close(p[RUNCAP_STDERR][0]);
		} else if (rc->rc_cap[RUNCAP_STDERR].sc_storfd != -1) {
			dup2(rc->rc_cap[RUNCAP_STDERR].sc_storfd,
			     RUNCAP_STDERR);
		}			
		
		i = open("/dev/null", O_RDONLY);
		if (i == -1)
			i = sysconf(_SC_OPEN_MAX) - 1;
		while (i > RUNCAP_STDERR) {
			close(i);
			i--;
		}

		if (rc->rc_env)
			environ = rc->rc_env;
		execvp(rc->rc_program ? rc->rc_program : rc->rc_argv[0],
		       rc->rc_argv);
		_exit(127);

		/* Parent branches */
	case -1:
		break;
		
	default:
		if (p[RUNCAP_STDIN][0] >= 0) {
			close(p[RUNCAP_STDIN][0]);
			rc->rc_cap[RUNCAP_STDIN].sc_fd = p[RUNCAP_STDIN][1];
		}
		if (p[RUNCAP_STDOUT][0] >= 0) {
			close(p[RUNCAP_STDOUT][1]);
			rc->rc_cap[RUNCAP_STDOUT].sc_fd = p[RUNCAP_STDOUT][0];
		}
		if (p[RUNCAP_STDERR][0] >= 0) {
			close(p[RUNCAP_STDERR][1]);
			rc->rc_cap[RUNCAP_STDERR].sc_fd = p[RUNCAP_STDERR][0];
		}
		return 0;
	}
  err:
	rc->rc_errno = errno;
	for (i = 0; i < RUNCAP_NBUF; i++) {
		close(p[i][0]);
		close(p[i][1]);
	}
	return -1;
}

static void
runcap_loop(struct runcap *rc)
{
	int nfd;
	fd_set rds, wrs;
	struct timeval finish, tv, *tvp;
		
	gettimeofday(&finish, NULL);
	finish.tv_sec += rc->rc_timeout;

	while (1) {
		int nready;
		int eof;

		nfd = -1;
		FD_ZERO(&rds);
		FD_ZERO(&wrs);

		if (rc->rc_cap[RUNCAP_STDIN].sc_size
		    && rc->rc_cap[RUNCAP_STDIN].sc_fd >= 0) {
			nfd = rc->rc_cap[RUNCAP_STDIN].sc_fd;
			FD_SET(rc->rc_cap[RUNCAP_STDIN].sc_fd, &wrs);
		}
		if (rc->rc_cap[RUNCAP_STDOUT].sc_fd >= 0) {
			if (rc->rc_cap[RUNCAP_STDOUT].sc_fd > nfd)
				nfd = rc->rc_cap[RUNCAP_STDOUT].sc_fd;
			FD_SET(rc->rc_cap[RUNCAP_STDOUT].sc_fd, &rds);
		}
		if (rc->rc_cap[RUNCAP_STDERR].sc_fd >= 0) {
			if (rc->rc_cap[RUNCAP_STDERR].sc_fd > nfd)
				nfd = rc->rc_cap[RUNCAP_STDERR].sc_fd;
			FD_SET(rc->rc_cap[RUNCAP_STDERR].sc_fd, &rds);
		}
		nfd++;

		if (rc->rc_pid != (pid_t) -1) {
			int flags = WNOHANG;
			pid_t pid;

			if (nfd == 0 && rc->rc_timeout == 0)
				flags = 0;
			pid = waitpid(rc->rc_pid, &rc->rc_status, flags);
			if (pid == -1) {
			        if (errno == EINTR)
				      continue;
				rc->rc_errno = errno;
				kill(rc->rc_pid, SIGKILL);
				break;
			}
			if (pid == rc->rc_pid)
				rc->rc_pid = (pid_t) -1;
		}

		if (nfd == 0 && rc->rc_pid == (pid_t) -1)
			break;
		
		if (rc->rc_timeout) {
			struct timeval now;
			gettimeofday(&now, NULL);
			tv = timeval_diff(&finish, &now);
			if (!timeval_after(&now, &finish)) {
				if (rc->rc_pid == (time_t) -1)
					break;
				kill(rc->rc_pid, SIGKILL);
				rc->rc_errno = ETIMEDOUT;
				continue;
			}
			tvp = &tv;
		} else {
			tvp = NULL;
		}

		nready = select(nfd, &rds, &wrs, NULL, tvp);
		if (nready == 0) {
			if (rc->rc_status)
				break;
			continue;
		}
		if (nready == -1) {
			if (errno == EINTR || errno == EAGAIN)
				/* retry */;
			else {
				rc->rc_errno = errno;
				break;
			}
			continue;
		}

		if (rc->rc_cap[RUNCAP_STDIN].sc_fd >= 0
		    && FD_ISSET(rc->rc_cap[RUNCAP_STDIN].sc_fd, &wrs)) {
			if (stream_capture_put(&rc->rc_cap[RUNCAP_STDIN], &eof)) {
				rc->rc_errno = errno;
				break;
			}
			if (eof) {
				/* FIXME: */
				close(rc->rc_cap[RUNCAP_STDIN].sc_fd);
				rc->rc_cap[RUNCAP_STDIN].sc_fd = -1;
			}
		}
		
		if (rc->rc_cap[RUNCAP_STDOUT].sc_fd >= 0
		    && FD_ISSET(rc->rc_cap[RUNCAP_STDOUT].sc_fd, &rds)) {
			if (stream_capture_get(&rc->rc_cap[RUNCAP_STDOUT], &eof)) {
				rc->rc_errno = errno;
				break;
			}
			if (eof) {
				close(rc->rc_cap[RUNCAP_STDOUT].sc_fd);
				rc->rc_cap[RUNCAP_STDOUT].sc_fd = -1;
			}
		}
		if (rc->rc_cap[RUNCAP_STDERR].sc_fd >= 0
		    && FD_ISSET(rc->rc_cap[RUNCAP_STDERR].sc_fd, &rds)) {
			if (stream_capture_get(&rc->rc_cap[RUNCAP_STDERR], &eof)) {
				rc->rc_errno = errno;
				break;
			}
			if (eof) {
				close(rc->rc_cap[RUNCAP_STDERR].sc_fd);
				rc->rc_cap[RUNCAP_STDERR].sc_fd = -1;
			}
		}
	}

	if (rc->rc_pid != (pid_t)-1) {
		kill(rc->rc_pid, SIGKILL);
		waitpid(rc->rc_pid, &rc->rc_status, 0);
		rc->rc_pid = (pid_t) -1;
	}
}

static int
runcap_init(struct runcap *rc, int flags)
{
	int res;
	
	if (!(flags & RCF_PROGRAM))
		rc->rc_program = NULL;
	if (!(flags & RCF_TIMEOUT))
		rc->rc_timeout = 0;
	if (!(flags & RCF_ENV))
		rc->rc_env = NULL;
	if (flags & RCF_STDIN) {
		if (rc->rc_cap[RUNCAP_STDIN].sc_size > 0
		    && rc->rc_cap[RUNCAP_STDIN].sc_fd != -1) {
			errno = EINVAL;
			return -1;
		}
		rc->rc_cap[RUNCAP_STDIN].sc_level =
			rc->rc_cap[RUNCAP_STDIN].sc_size;
		rc->rc_cap[RUNCAP_STDIN].sc_cur = 0;
		rc->rc_cap[RUNCAP_STDIN].sc_storfd = -1;
	} else if (stream_capture_init(&rc->rc_cap[RUNCAP_STDIN], 0, 0))
		return -1;

	res = stream_capture_init(&rc->rc_cap[RUNCAP_STDOUT],
				  (flags & RCF_STDOUT_SIZE)
				    ? rc->rc_cap[RUNCAP_STDOUT].sc_size
				    : STRCAP_BUFSIZE,
				  RCF_FLAG_TO_SC(flags, RUNCAP_STDOUT));
	if (res)
		return res;
	
	res = stream_capture_init(&rc->rc_cap[RUNCAP_STDERR],
				  (flags & RCF_STDERR_SIZE)
				    ? rc->rc_cap[RUNCAP_STDERR].sc_size
				    : STRCAP_BUFSIZE,
				  RCF_FLAG_TO_SC(flags, RUNCAP_STDERR));
	if (res)
		return res;

	rc->rc_pid = (pid_t) -1;
	rc->rc_status = 0;
	rc->rc_errno = 0;

	return 0;
}

static void
set_signals(void (*handler)(int signo), int sigc, int *sigv,
	    struct sigaction *oldact)
{
	int i;
	struct sigaction act;

	act.sa_flags = 0;
	sigemptyset(&act.sa_mask);
	for (i = 0; i < sigc; i++)
		sigaddset(&act.sa_mask, i);

	for (i = 0; i < sigc; i++) {
		act.sa_handler = handler;
		sigaction(sigv[i], &act, &oldact[i]);
	}
}

static void
restore_signals(int sigc, int *sigv, struct sigaction *oldact)
{
	int i;
	for (i = 0; i < sigc; i++)
		sigaction(sigv[i], &oldact[i], NULL);
}	

static void
sighan(int signo)
{
	/* nothing */
}

int
runcap(struct runcap *rc, int flags)
{
	int sig[] = { SIGCHLD, SIGPIPE };
#define NUMSIGNALS (sizeof(sig)/sizeof(sig[0]))
	struct sigaction oldact[NUMSIGNALS];
	
	if (runcap_init(rc, flags)) {
		rc->rc_errno = errno;
		return -1;
	}
	
	set_signals(sighan, NUMSIGNALS, sig, oldact);
	if (runcap_start(rc) == 0)
		runcap_loop(rc);
	restore_signals(NUMSIGNALS, sig, oldact);
	if (rc->rc_errno == 0) {
		if (rc->rc_cap[RUNCAP_STDOUT].sc_storfd != -1) {
			stream_capture_flush(&rc->rc_cap[RUNCAP_STDOUT]);
			lseek(rc->rc_cap[RUNCAP_STDOUT].sc_storfd, 0, SEEK_SET);
		}
		if (rc->rc_cap[RUNCAP_STDERR].sc_storfd != -1) {
			stream_capture_flush(&rc->rc_cap[RUNCAP_STDERR]);
			lseek(rc->rc_cap[RUNCAP_STDERR].sc_storfd, 0, SEEK_SET);
		}
	}
	if (rc->rc_errno) {
		errno = rc->rc_errno;
		return -1;
	}
	runcap_rewind(rc, RUNCAP_STDOUT);
	runcap_rewind(rc, RUNCAP_STDERR);
	return 0;
}
	

